File: java-backend/src/test/java/com/study/service/AuthServiceTest.java
package com.study.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

import com.study.domain.User;
import com.study.exception.ConflictException;
import com.study.exception.UnauthorizedException;
import com.study.mapper.UserMapper;
import com.study.security.JwtService;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.crypto.password.PasswordEncoder;

/** AuthService 单元测试(纯 Mockito,mock UserMapper)。 */
@ExtendWith(MockitoExtension.class)
class AuthServiceTest {

    @Mock
    UserMapper userMapper;
    @Mock
    PasswordEncoder passwordEncoder;
    @Mock
    JwtService jwtService;
    @InjectMocks
    AuthService authService;

    @Test
    void register_success_encodesPasswordAndInserts() {
        when(userMapper.selectCount(any())).thenReturn(0L);
        when(passwordEncoder.encode("pw")).thenReturn("hash");
        when(userMapper.insert(any(User.class))).thenReturn(1);

        User user = authService.register("alice", "pw");

        assertThat(user.getUsername()).isEqualTo("alice");
        assertThat(user.getPasswordHash()).isEqualTo("hash");
    }

    @Test
    void register_duplicateUsername_throwsConflict() {
        when(userMapper.selectCount(any())).thenReturn(1L);

        assertThatThrownBy(() -> authService.register("alice", "pw"))
                .isInstanceOf(ConflictException.class);
    }

    @Test
    void login_validCredentials_returnsToken() {
        User user = new User();
        user.setId(1L);
        user.setUsername("alice");
        user.setPasswordHash("hash");
        when(userMapper.selectOne(any())).thenReturn(user);
        when(passwordEncoder.matches("pw", "hash")).thenReturn(true);
        when(jwtService.generateToken(1L, "alice")).thenReturn("tok");

        assertThat(authService.login("alice", "pw")).isEqualTo("tok");
    }

    @Test
    void login_wrongPassword_throwsUnauthorized() {
        User user = new User();
        user.setUsername("alice");
        user.setPasswordHash("hash");
        when(userMapper.selectOne(any())).thenReturn(user);
        when(passwordEncoder.matches("bad", "hash")).thenReturn(false);

        assertThatThrownBy(() -> authService.login("alice", "bad"))
                .isInstanceOf(UnauthorizedException.class);
    }

    @Test
    void login_unknownUser_throwsUnauthorized() {
        when(userMapper.selectOne(any())).thenReturn(null);

        assertThatThrownBy(() -> authService.login("ghost", "pw"))
                .isInstanceOf(UnauthorizedException.class);
    }
}

File: kotlin-backend/src/test/kotlin/com/study/service/AuthServiceTest.kt
package com.study.service

import com.baomidou.mybatisplus.core.conditions.Wrapper
import com.study.domain.User
import com.study.exception.ConflictException
import com.study.exception.UnauthorizedException
import com.study.mapper.UserMapper
import com.study.security.JwtService
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.springframework.security.crypto.password.PasswordEncoder

/** AuthService 单元测试(MockK,mock UserMapper)。 */
class AuthServiceTest {

    private val userMapper = mockk<UserMapper>()
    private val passwordEncoder = mockk<PasswordEncoder>()
    private val jwtService = mockk<JwtService>()
    private val authService = AuthService(userMapper, passwordEncoder, jwtService)

    @Test
    fun `注册成功 加密并插入`() {
        every { userMapper.selectCount(any<Wrapper<User>>()) } returns 0L
        every { passwordEncoder.encode("pw") } returns "hash"
        every { userMapper.insert(any<User>()) } returns 1

        val user = authService.register("alice", "pw")

        assertEquals("alice", user.username)
        assertEquals("hash", user.passwordHash)
    }

    @Test
    fun `用户名重复 抛冲突`() {
        every { userMapper.selectCount(any<Wrapper<User>>()) } returns 1L

        assertThrows(ConflictException::class.java) { authService.register("alice", "pw") }
    }

    @Test
    fun `登陆成功 返回令牌`() {
        val user = User().apply {
            id = 1L
            username = "alice"
            passwordHash = "hash"
        }
        every { userMapper.selectOne(any<Wrapper<User>>()) } returns user
        every { passwordEncoder.matches("pw", "hash") } returns true
        every { jwtService.generateToken(1L, "alice") } returns "tok"

        assertEquals("tok", authService.login("alice", "pw"))
    }

    @Test
    fun `密码错误 抛未授权`() {
        val user = User().apply {
            username = "alice"
            passwordHash = "hash"
        }
        every { userMapper.selectOne(any<Wrapper<User>>()) } returns user
        every { passwordEncoder.matches("bad", "hash") } returns false

        assertThrows(UnauthorizedException::class.java) { authService.login("alice", "bad") }
    }

    @Test
    fun `用户不存在 抛未授权`() {
        every { userMapper.selectOne(any<Wrapper<User>>()) } returns null

        assertThrows(UnauthorizedException::class.java) { authService.login("ghost", "pw") }
    }
}
